File: MATLAB_GA_Algorithm/GeneticAlgorithm.m
clc;
clear;
close all;


%% Problem Definition

problem.CostFunction = @(x) Sphere(x);      % Cost function
problem.nVar = 6;                           % number of changing variables
problem.VarMin = [1 -1 1 1 -1 1];         % Upper bound
problem.VarMax = [10 10 10 10 10 10];             % Lower bound
problem.FindMin = true;                    % True for find minimum, false for find maximum


%% GA Parameters

params.MaxIt = 500;                         % Maximum number of GA iterations
params.nPop = 200;                           % Population size

params.beta = 1;                            % Parent score coefficient (<0 for find the minimum, >0 for find the maximum)
params.c_percentage = 1;                    % Number of offspring compare to Population size
params.mutation_rate = 0.001;               % Mutation rate
params.sigma = 0.1;                         % standard deviation of the gaussian noise in the mutation

params.ShowIterInfo = true;                 % Show interation

%% Run GA

result = GA(problem, params);


%% Result

figure;
plot(result.BestCosts_List, 'LineWidth', 2);
title('Convergent plot (GA)');
xlabel('Iteration');
ylabel('Best Cost');
grid on;

File: Python_GA_Algorithm/GeneticAlgorithm.py
import numpy as np
import matplotlib.pyplot as plt
from ToolClasses import ProblemClass, ParamsClass           # import tools
from CostFunction import *                                  # import cost function
from GA_Function import *                                   # import GA

################################################################################################
# Problem Definition
problem = ProblemClass()
problem.CostFunction = sphere
problem.nVar = 5
problem.VarMin = [-1, 8, -1, -1, -1]
problem.VarMax = [10, 10, 10, 10, 10]
problem.FindMin = True

################################################################################################
# GA Parameters
params = ParamsClass()
params.MaxIt = 400
params.nPop = 200
params.beta = 1
params.offspring_percentage = 1
params.Mutation_rate = 0.001
params.ShowIterInfo = True

################################################################################################
# Run GA
result = GA(problem, params)
print("Best Cost = {}".format(result.GlobalBest.cost))
print("Best Solution = {}".format(result.GlobalBest.position))

################################################################################################
# Results
plt.plot(result.BestCosts_List)
# plt.semilogy(result.BestCosts_List)
plt.xlim(0, params.MaxIt)
plt.xlabel('Iterations')
plt.ylabel('Best Cost')
if problem.FindMin:
    plt.title('(GA) Minimum Convergent Plot')
else:
    plt.title('(GA) Maximum Convergent Plot')
plt.grid(True)

plt.show()